### PR TITLE
Enrich mock api

### PR DESCRIPTION
**Pull request by rlecomte**
## What this PR does

Support of Query parameters, Header and Body input in mock configuration.

TODO: filtering isn't implemented yet. (see #61)

## Related

See: #56 #61

### CODE REVIEW

File: dhall/Mock/package.dhall [hunk 1]
```diff
@@ -1,23 +1,67 @@
...
+  { mapKey = "Content-Type", mapValue = "text/plain"}
+
+in { HttpMethod      = HttpMethod
+   , QueryParam      = QueryParam
+   , Header          = Header
+   , Body            = Body
+   , HttpRequest     = HttpRequest
+   , HttpResponse    = HttpResponse
+   , Expectation     = Expectation
+   , statusOK            = Some 200
```
Comment (fteychene) [inline dhall/Mock/package.dhall:58]: Could we add all the status in a custom object ?
Something like 
```
Status = {
  OK = Some 200
  , Created = Some 201 
  , ...
}
```

File: dhall/static.dhall [hunk 2]
```diff
@@ -1,21 +1,19 @@
...
-                                   , path    = Some "/greet/pwet"
-                                   }
-                       , response = { statusCode   = Some +201
-                                       , statusReason = None Text
-                                       , body         = Some "Hello, pwet ! Comment que ca biche ?"
-                                       }
+                       { request  = Mock.HttpRequest::{ method  = Some Mock.HttpMethod.GET
+                                                      , path    = Some "/greet/pwet"
+                                                      }
+                       , response = Mock.HttpResponse::{ statusCode   = Some 201
```
Comment (fteychene) [inline dhall/static.dhall:7]: Use status defined in mock ?

File: dhall/static.dhall [hunk 3]
```diff
@@ -1,21 +1,19 @@
...
-                                   , path    = Some "/greet/wololo"
-                                   }
-                      , response = { statusCode   = Some +200
-                                   , statusReason = None Text
-                                   , body         = Some "Hello, Wololo !"
-                                   }
+                      ,{ request  = Mock.HttpRequest::{ method  = Some Mock.HttpMethod.GET
+                                                      , path    = Some "/greet/wololo"
+                                                      }
+                      , response = Mock.HttpResponse::{ statusCode   = Some 200
```
Comment (fteychene) [inline dhall/static.dhall:14]: Use status defined in mock ?

File: src/mock/compilation.rs [hunk 4]
```diff
@@ -11,21 +11,20 @@ pub fn compile_configuration(configuration_content: &str) -> Result<Vec<Expectat
 mod test {
     use super::*;
     use crate::mock::model::{Expectation, HttpMethod, HttpRequest, HttpResponse};
+    use std::collections::BTreeMap;
```
Comment (fteychene) [inline src/mock/compilation.rs:14]: Why could we not use a standard HashMap ?

File: src/mock/serde/mod.rs [hunk 5]
```diff
@@ -0,0 +1,45 @@
+pub mod json_string {
```
Comment (fteychene) [inline src/mock/serde/mod.rs:1]: Why do we define this module ?
It seems that it is not used currently ?

File: src/mock/compilation.rs (hunk 4 above)
Comment (rlecomte) [reply src/mock/compilation.rs:14]: We could, the memory footprint will be higher

File: src/mock/serde/mod.rs (hunk 5 above)
Comment (rlecomte) [reply src/mock/serde/mod.rs:1]: It's used in `mock/model.rs`  : 
```rust
#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
pub enum RequestBody {
    JSON {
        #[serde(with = "serde_mock::json_string")]
        json: Value,
    },
    TEXT {
        text: String,
    },
}
```

File: src/mock/compilation.rs (hunk 4 above)
Comment (rlecomte) [reply src/mock/compilation.rs:14]: But you're right hasmap is a better choice, I will change that tomorrow

File: dhall/Mock/package.dhall (hunk 1 above)
Comment (rlecomte) [reply dhall/Mock/package.dhall:58]: good idea

Comment (rlecomte) [reply dhall/Mock/package.dhall:58]: In convention in dhall, only Type start with an uppercase, but we can get rid of this. What do you think?

Comment (rlecomte) [reply dhall/Mock/package.dhall:58]: It's also shorter to define status by :

```
Mock.statusOK
```

than : 
```
Mock.Status.OK
```

File: dhall/static.dhall (hunk 2 above)
Comment (rlecomte) [reply dhall/static.dhall]: done

File: dhall/static.dhall (hunk 3 above)
Comment (rlecomte) [reply dhall/static.dhall]: done

File: src/mock/compilation.rs (hunk 4 above)
Comment (rlecomte) [reply src/mock/compilation.rs]: done

File: dhall/Mock/package.dhall (hunk 1 above)
Comment (fteychene) [reply dhall/Mock/package.dhall:58]: OK Tas more thinking about doing an import on Status direct but fine with me

**Pull request merged**